python/cudf_polars/tests/dsl/test_traversal.py

```diff
@@ -0,0 +1,163 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import pylibcudf as plc
+
+import polars as pl
+from polars.testing.asserts import assert_frame_equal
+
+from cudf_polars import translate_ir
+from cudf_polars.dsl import expr, ir
+from cudf_polars.dsl.traversal import (
+    CachingVisitor,
+    make_recursive,
+    reuse_if_unchanged,
+    traversal,
+)
+
+
+def make_expr(dt, n1, n2):
+    a1 = expr.Col(dt, n1)
+    a2 = expr.Col(dt, n2)
+
+    return expr.BinOp(dt, plc.binaryop.BinaryOperator.MUL, a1, a2)
+
+
+def test_traversal_unique():
+    dt = plc.DataType(plc.TypeId.INT8)
+
+    e1 = make_expr(dt, "a", "a")
+    unique_exprs = list(traversal(e1))
+
+    assert len(unique_exprs) == 2
+    assert set(unique_exprs) == {expr.Col(dt, "a"), e1}
+    assert unique_exprs == [e1, expr.Col(dt, "a")]
+
+    e2 = make_expr(dt, "a", "b")
+    unique_exprs = list(traversal(e2))
+
+    assert len(unique_exprs) == 3
+    assert set(unique_exprs) == {expr.Col(dt, "a"), expr.Col(dt, "b"), e2}
+    assert unique_exprs == [e2, expr.Col(dt, "a"), expr.Col(dt, "b")]
+
+    e3 = make_expr(dt, "b", "a")
+    unique_exprs = list(traversal(e3))
+
+    assert len(unique_exprs) == 3
+    assert set(unique_exprs) == {expr.Col(dt, "a"), expr.Col(dt, "b"), e3}
+    assert unique_exprs == [e3, expr.Col(dt, "b"), expr.Col(dt, "a")]
+
+
+def rename(e, rec):
+    if isinstance(e, expr.Col) and e.name in rec.mapping:
+        return type(e)(e.dtype, rec.mapping[e.name])
```

Comment:
Note one can't use `e.reconstruct` here, because these attributes are not children.

Reply:
Just to check my understanding: The purpose of `reconstruct` is to "re-plumb" the same node type and non-child options to different children. It is not meant to substitute a node with a different type (or the same type with different non-child options).

Reply:
Yes, exactly. This renaming visitor is a typical example. Suppose we have:

```
e = BinOp(dtype, ADD, Col("a"), Col("b"))
```
And we want to replace `Col("a")` by an equivalent expression referencing a different column. So we dive down and see `Col("a")` and produce a new name (as above), say `Col("c")`. Then when we come back up, we want to remake the `BinOp` with the new object, but leave everything else unchanged (the operator hasn't changed), so we would do `e.reconstruct(Col("c"), Col("b"))` (morally).

The `reuse_if_unchanged` recipe is a bit smarter, assuming that what you always want to do is recurse to the leaves and rebuild from there, it does the recursion onto the children for you and then calls reconstruct with the new children. _But_ it only does so if the children have changed. It turns out if you have big expressions this can save you a lot, because suppose the part of the BinOp you didn't rename a column in was something large, then you don't pay the price of rebuilding anything.